Improving plantuml generation to support:

robwiss opened the pull request:
* on_entry and on_exit
* boolean operators for guards !, &&, and ||
* multiple actions
* orthogonal states
* submachines
* replace "terminate" state with plantuml termination syntax [*]
* update example to show these changes + use functor style instead of lambda style so the generated plantuml reads nicer

I strongly suspect this PR will get rejected because it uses C++17 instead of C++14. However, the improvements to plantuml generation are substantial. Is there a way this example could be included for those who use C++17 or could help be provided in removing the reliance on `if constexpr`?

Thank you.

krzysztof-jusiak commented:
nice @robwiss

krzysztof-jusiak commented on Makefile:
```diff
@@ -6,7 +6,7 @@
 .PHONY: all doc clean test example
 CXX?=clang++
-CXXSTD?=c++14
+CXXSTD?=c++17
```
as long as I would like to, I don't we can change it globally just like that, can we either make the plant_uml example c++17 only or make it c++14 compatible

robwiss replied:
I'll look at doing that. Perhaps it can be feature flagged to use the old example in C++14 mode and this one in C++17.

krzysztof-jusiak commented on include/boost/sml/aux_/utility.hpp:
```diff
@@ -152,7 +152,8 @@ struct pool_type_impl : pool_type_base {
 };
 
 template <class T>
-struct pool_type_impl<T &, aux::enable_if_t<aux::is_constructible<T>::value && aux::is_constructible<T, T>::value>> : pool_type_base {
```
is that clang.format > 9? ci is tseting against clang-format 9 I think

robwiss replied:
It was clang-format-10. I switched to clang-format-9 and re-ran. Hopefully it's good now.

robwiss commented:
I forgot to mention another feature of this approach. It is not required to instantiate the state machine type to dump the plantuml. Great for build times!

krzysztof-jusiak commented:
> I forgot to mention another feature of this approach. It is not required to instantiate the state machine type to dump the plantuml. Great for build times!

Very nice :+1:

robwiss commented:
I changed `plant_uml.cpp` to compile the more featured version when C++17 or greater is available and the existing version when C++14 is available.

I added an `example/plant_uml.out` rule to the Makefile. It attempts to build with `-std=c++17` then falls back to `-std=c++14` if that fails.
